File: dev-portal/typescript/src/services/ResetService.ts
/**
 * Reset Service
 * 
 * Full workspace reset workflow
 */

import { PostmanClient } from '../client';
import {
  Workspace,
  Collection,
  Environment,
  MockServer,
  Spec,
  ProgressCallback,
} from '../types';

export interface ResetConfig {
  client: PostmanClient;
  workspaceId: string;
  onProgress?: ProgressCallback;
}

export interface ResetResult {
  workspace: Workspace | null;
  specs: {
    total: number;
    deleted: number;
    failed: Array<{ name: string; error: string }>;
  };
  mocks: {
    total: number;
    deleted: number;
    failed: Array<{ name: string; error: string }>;
  };
  environments: {
    total: number;
    deleted: number;
    failed: Array<{ name: string; error: string }>;
  };
  collections: {
    total: number;
    deleted: number;
    failed: Array<{ name: string; error: string }>;
  };
}

export interface WorkspaceContents {
  collections: Collection[];
  environments: Environment[];
  mocks: MockServer[];
  specs: Spec[];
}

/**
 * Reset Service for workspace cleanup
 */
export class ResetService {
  private client: PostmanClient;
  private config: ResetConfig;

  constructor(config: ResetConfig) {
    this.client = config.client;
    this.config = config;
  }

  /**
   * Scan workspace to get all contents
   */
  async scanWorkspace(): Promise<{
    workspace: Workspace | null;
    contents: WorkspaceContents;
    total: number;
  }> {
    const [workspace, collections, environments, mocks, specs] = await Promise.all([
      this.client.getWorkspace(this.config.workspaceId),
      this.client.getCollections(this.config.workspaceId),
      this.client.getEnvironments(this.config.workspaceId),
      this.client.getMocks(this.config.workspaceId),
      this.client.getSpecs(this.config.workspaceId),
    ]);

    return {
      workspace,
      contents: { collections, environments, mocks, specs },
      total: collections.length + environments.length + mocks.length + specs.length,
    };
  }

  /**
   * Run full reset workflow
   * 
   * Deletion order (reverse of provisioning):
   * 1. Specs first
   * 2. Mocks (depend on collections)
   * 3. Environments
   * 4. Collections last
   */
  async reset(): Promise<ResetResult> {
    const result = this.initializeResult();

    // Scan workspace
    this.emitProgress('scan', 'Scanning workspace contents...');
    const { workspace, contents, total } = await this.scanWorkspace();
    result.workspace = workspace;

    if (total === 0) {
      this.emitProgress('complete', 'Workspace is already empty');
      return result;
    }

    result.specs.total = contents.specs.length;
    result.mocks.total = contents.mocks.length;
    result.environments.total = contents.environments.length;
    result.collections.total = contents.collections.length;

    // Step 1: Delete specs
    this.emitProgress('specs', `Deleting ${contents.specs.length} spec(s)...`);
    await this.deleteSpecs(contents.specs, result);

    // Step 2: Delete mocks
    this.emitProgress('mocks', `Deleting ${contents.mocks.length} mock server(s)...`);
    await this.deleteMocks(contents.mocks, result);

    // Step 3: Delete environments
    this.emitProgress('environments', `Deleting ${contents.environments.length} environment(s)...`);
    await this.deleteEnvironments(contents.environments, result);

    // Step 4: Delete collections
    this.emitProgress('collections', `Deleting ${contents.collections.length} collection(s)...`);
    await this.deleteCollections(contents.collections, result);

    this.emitProgress('complete', 'Reset complete');
    return result;
  }

  private initializeResult(): ResetResult {
    return {
      workspace: null,
      specs: { total: 0, deleted: 0, failed: [] },
      mocks: { total: 0, deleted: 0, failed: [] },
      environments: { total: 0, deleted: 0, failed: [] },
      collections: { total: 0, deleted: 0, failed: [] },
    };
  }

  private emitProgress(step: string, message: string): void {
    this.config.onProgress?.({ step, message });
  }

  private async deleteSpecs(specs: Spec[], result: ResetResult): Promise<void> {
    for (const spec of specs) {
      const deleteResult = await this.client.deleteSpec(spec.id);

      if (deleteResult.success) {
        result.specs.deleted++;
      } else {
        result.specs.failed.push({ name: spec.name, error: deleteResult.error ?? 'Unknown error' });
      }

      await this.delay(300);
    }
  }

  private async deleteMocks(mocks: MockServer[], result: ResetResult): Promise<void> {
    for (const mock of mocks) {
      // Use mock.id (not mock.uid) for deletion
      const success = await this.client.deleteMock(mock.id);

      if (success) {
        result.mocks.deleted++;
      } else {
        result.mocks.failed.push({ name: mock.name, error: 'Failed to delete' });
      }

      await this.delay(300);
    }
  }

  private async deleteEnvironments(environments: Environment[], result: ResetResult): Promise<void> {
    for (const env of environments) {
      const success = await this.client.deleteEnvironment(env.uid);

      if (success) {
        result.environments.deleted++;
      } else {
        result.environments.failed.push({ name: env.name, error: 'Failed to delete' });
      }

      await this.delay(300);
    }
  }

  private async deleteCollections(collections: Collection[], result: ResetResult): Promise<void> {
    for (const collection of collections) {
      const success = await this.client.deleteCollection(collection.uid);

      if (success) {
        result.collections.deleted++;
      } else {
        result.collections.failed.push({ name: collection.name, error: 'Failed to delete' });
      }

      await this.delay(300);
    }
  }

  private delay(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: dev-portal/python/postman_sdk/services/reset_service.py
"""
Reset Service

Full workspace reset workflow
"""

import asyncio
from typing import Any, Callable

from postman_sdk.client import PostmanClient
from postman_sdk.types import (
    Workspace,
    Collection,
    Environment,
    MockServer,
    Spec,
    ProgressEvent,
)

ProgressCallback = Callable[[ProgressEvent], None]


class ResetService:
    """Reset Service for workspace cleanup"""

    def __init__(
        self,
        client: PostmanClient,
        workspace_id: str,
        on_progress: ProgressCallback | None = None,
    ):
        self.client = client
        self.workspace_id = workspace_id
        self.on_progress = on_progress

    async def scan_workspace(self) -> dict[str, Any]:
        """Scan workspace to get all contents"""
        workspace, collections, environments, mocks, specs = await asyncio.gather(
            self.client.get_workspace(self.workspace_id),
            self.client.get_collections(self.workspace_id),
            self.client.get_environments(self.workspace_id),
            self.client.get_mocks(self.workspace_id),
            self.client.get_specs(self.workspace_id),
        )

        return {
            "workspace": workspace,
            "contents": {
                "collections": collections,
                "environments": environments,
                "mocks": mocks,
                "specs": specs,
            },
            "total": len(collections) + len(environments) + len(mocks) + len(specs),
        }

    async def reset(self) -> dict[str, Any]:
        """
        Run full reset workflow

        Deletion order (reverse of provisioning):
        1. Specs first
        2. Mocks (depend on collections)
        3. Environments
        4. Collections last
        """
        result = self._initialize_result()

        # Scan workspace
        self._emit_progress("scan", "Scanning workspace contents...")
        scan_result = await self.scan_workspace()
        result["workspace"] = scan_result["workspace"]
        contents = scan_result["contents"]

        if scan_result["total"] == 0:
            self._emit_progress("complete", "Workspace is already empty")
            return result

        result["specs"]["total"] = len(contents["specs"])
        result["mocks"]["total"] = len(contents["mocks"])
        result["environments"]["total"] = len(contents["environments"])
        result["collections"]["total"] = len(contents["collections"])

        # Step 1: Delete specs
        self._emit_progress("specs", f"Deleting {len(contents['specs'])} spec(s)...")
        await self._delete_specs(contents["specs"], result)

        # Step 2: Delete mocks
        self._emit_progress("mocks", f"Deleting {len(contents['mocks'])} mock server(s)...")
        await self._delete_mocks(contents["mocks"], result)

        # Step 3: Delete environments
        self._emit_progress("environments", f"Deleting {len(contents['environments'])} environment(s)...")
        await self._delete_environments(contents["environments"], result)

        # Step 4: Delete collections
        self._emit_progress("collections", f"Deleting {len(contents['collections'])} collection(s)...")
        await self._delete_collections(contents["collections"], result)

        self._emit_progress("complete", "Reset complete")
        return result

    def _initialize_result(self) -> dict[str, Any]:
        return {
            "workspace": None,
            "specs": {"total": 0, "deleted": 0, "failed": []},
            "mocks": {"total": 0, "deleted": 0, "failed": []},
            "environments": {"total": 0, "deleted": 0, "failed": []},
            "collections": {"total": 0, "deleted": 0, "failed": []},
        }

    def _emit_progress(self, step: str, message: str) -> None:
        if self.on_progress:
            self.on_progress(ProgressEvent(step=step, message=message))

    async def _delete_specs(self, specs: list[Spec], result: dict[str, Any]) -> None:
        for spec in specs:
            delete_result = await self.client.delete_spec(spec.id)

            if delete_result.success:
                result["specs"]["deleted"] += 1
            else:
                result["specs"]["failed"].append({
                    "name": spec.name,
                    "error": delete_result.error or "Unknown error",
                })

            await asyncio.sleep(0.3)

    async def _delete_mocks(self, mocks: list[MockServer], result: dict[str, Any]) -> None:
        for mock in mocks:
            # Use mock.id (not mock.uid) for deletion
            success = await self.client.delete_mock(mock.id)

            if success:
                result["mocks"]["deleted"] += 1
            else:
                result["mocks"]["failed"].append({
                    "name": mock.name,
                    "error": "Failed to delete",
                })

            await asyncio.sleep(0.3)

    async def _delete_environments(self, environments: list[Environment], result: dict[str, Any]) -> None:
        for env in environments:
            success = await self.client.delete_environment(env.uid)

            if success:
                result["environments"]["deleted"] += 1
            else:
                result["environments"]["failed"].append({
                    "name": env.name,
                    "error": "Failed to delete",
                })

            await asyncio.sleep(0.3)

    async def _delete_collections(self, collections: list[Collection], result: dict[str, Any]) -> None:
        for collection in collections:
            success = await self.client.delete_collection(collection.uid)

            if success:
                result["collections"]["deleted"] += 1
            else:
                result["collections"]["failed"].append({
                    "name": collection.name,
                    "error": "Failed to delete",
                })

            await asyncio.sleep(0.3)
